### DISCUSSION TITLE
Dagger Function "initialize" times are prohibitively long (0.11.0)

### DISCUSSION THREAD

**Discussion by verdverm** [dagger/dagger#12323] [Help]
### What is the issue?

The startup time for Dagger Functions takes a long time, even on subsequent runs. It seems to be the step where the function is prepared, compiled, dependencies fetched. Watching the logs, once the function code is reached, it goes very quickly

Timings are very inconsistent

- stark difference with TS (much longer)
- subsequent runs can take longer than the first (PY)
- 

### Dagger version

0.11.0

### Steps to reproduce

```
cd tmp 
➜  tmp git:(main) ✗  
dagger init --sdk=go repro-go
12:01:19 WRN no LICENSE file found; generating one for you, feel free to change or remove license=Apache-2.0
Initialized module repro-go in repro-go
➜  tmp git:(main) ✗  
dagger init --sdk=python repro-py
12:01:53 WRN no LICENSE file found; generating one for you, feel free to change or remove license=Apache-2.0
Initialized module repro-py in repro-py
➜  tmp git:(main) ✗  
dagger init --sdk=typescript repro-ts
12:02:26 WRN no LICENSE file found; generating one for you, feel free to change or remove license=Apache-2.0
Initialized module repro-ts in repro-ts
➜  tmp git:(main) ✗  
ls
repro-go repro-py repro-ts


➜  tmp git:(main) ✗  
cd repro-go 
➜  repro-go git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  6.55s user 3.33s system 25% cpu 38.000 total
➜  repro-go git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  3.40s user 1.79s system 69% cpu 7.474 total
➜  repro-go git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  7.11s user 3.53s system 26% cpu 39.969 total


➜  repro-go git:(main) ✗  
cd ../repro-py 
➜  repro-py git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  1.63s user 1.00s system 14% cpu 18.072 total
➜  repro-py git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  3.24s user 1.14s system 12% cpu 34.902 total


➜  repro-py git:(main) ✗  
cd ../repro-ts
➜  repro-ts git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  13.74s user 4.48s system 15% cpu 1:55.68 total
➜  repro-ts git:(main) ✗  
time dagger call container-echo --string-arg='Hello Daggernauts!' stdout
Hello Daggernauts!
dagger call container-echo --string-arg='Hello Daggernauts!' stdout  13.56s user 4.68s system 15% cpu 1:54.88 total
```

### Log output

(see above)

**Comment by TomChv** [comment]
Thanks for the report! We are aware of the init time and are actively working on it 🚀 

> stark difference with TS (much longer)

About the TypeScript, we just merged our support for `Bun` #6591 that can be used to divide by 2 to 3 the init time for TS modules.
It's still experimental but it's a conveniant way to improve performances if you need some.

**Comment by TomChv** [comment]
Duplicate of #7093, please follow the performance improvement on this issue instead

**Comment by helderco** [comment]
That issue is specific to TS while this one is not, so reopening.

**Comment by levlaz** [comment]
Hey @verdverm we've made some huge progress on performance over the last few releases, could you update to latest and re-run your benchmarks and let us know how it goes?

**Comment by verdverm** [comment]
When rerunning "hello world" after a first run:
- Go timings are the most variable, 2-12s
- Py times are around 3s
- Ts times are around 4s

For context, when I rebuild a complex Go application (100k LOC), it's 0.3s and the Py/TS runs I have feel instantaneous 

Average repro-go run:

```sh
time dagger -v call container-echo --string-arg='Hello Daggernauts!' stdout
✔ connect 0.7s
  ✔ starting engine 0.3s
  ✔ connecting to engine 0.1s
  ✔ starting session 0.2s
✔ initialize 4.8s
✔ ReproGo.containerEcho(stringArg: "Hello Daggernauts!"): Container! 0.4s
  ✔ Container.from(address: "alpine:latest"): Container! 0.2s

Hello Daggernauts!
dagger -v call container-echo --string-arg='Hello Daggernauts!' stdout  1.79s user 1.81s system 57% cpu 6.248 total
```

Another example that remains around 1.9s consistently

```
time dagger -v -m github.com/shykes/daggerverse/wolfi@v0.1.2 call container with-exec --args="cat","/etc/os-release" stdout
✔ connect 0.7s
  ✔ starting engine 0.3s
  ✔ connecting to engine 0.1s
  ✔ starting session 0.2s
✔ initialize 0.6s
✔ Wolfi.container: Container! 0.4s
  ✔ Apko.wolfi(packages: []): Container! 0.3s

ID=wolfi
NAME="Wolfi"
PRETTY_NAME="Wolfi"
VERSION_ID="20230201"
HOME_URL="https://wolfi.dev"
dagger -v -m github.com/shykes/daggerverse/wolfi@v0.1.2 call container     0.13s user 0.09s system 10% cpu 1.941 total
```

**Comment by levlaz** [comment]
Thanks so much for re-running, looks like we still got some work to do here, lets try again in the near future :)

**Comment by verdverm** [comment]
I would expect the `connect` phase to be nearly instantaneous, given the engine is already running, and I am running the same command in a loop

I would think a single API call ought to be enough to know it's running and in a good state

**Comment by levlaz** [comment]
Yeah definitely, in the general case we suggest not including
node_modules this way. Outside of the performance implications it can
also lead to issues between local/ci compatibility. For example if
you're working on MacOS locally but running Linux in CI. 

We're still working on adding this to our docs, but we do shows how to
exclude node_modules using "views" in our own dagger.json file here: https://github.com/dagger/dagger/blob/main/dagger.json#L24 

Once you have your view set up you then reference it like this: 

dagger call --src .:default

**Comment by TomChv** [comment]
Hey @verdverm did you try with v0.12? We tried to make Dagger significantly faster!

If that's not the case, let us know because we're doing our best to make it fast.
There's still several things happening on the TS SDK to make it faster, I might have unlock an optimization with https://github.com/dagger/dagger/pull/7864 that I'll do once it's merge, hopefully this will also reduce the initialization time.

**Comment by anton-slashm** [comment]
@TomChv for dagger `0.13.1` I also experience slow startup times of ~20 secs for python poetry project. `Installing module` and `analyzing module` are the culprits. It varies heavily depending, I think, on whether the code has changed or not, like there's a caching mechanism behind, but if I did change the module - I'll have to wait.

**Comment by helderco** [comment]
We do have some cache misses we're investigating:
- https://github.com/dagger/dagger/issues/8254

But it's likely that the module's uploading more than it needs to. I'm interested in whether this POC helps with your _Installing module_ phase (i.e., module load):
- https://github.com/dagger/dagger/pull/8501

For the _analyzing_ part, I didn't expect you to see an impact there. That phase is just doing module introspection in order to build up the CLI (later, in `prepare`). How long can it take for you in this step?

**Comment by attekei** [comment]
Experiencing 20s initialisation as well in a Python monorepo project. My gut feeling is that regardless of strict includes&excludes in dagger.json (btw that json lacks documentation), too much code gets still included. 

This is the step which is super slow:
```
moduleSource(refString: "[path-to-dagger-project]", relHostPath: "."): ModuleSource!
```

Tangential note: I'm additionally confused of how can I change the context where Dagger operates. It seems by default to use the git repository. 
* There seems to be https://docs.dagger.io/manuals/developer/functions/#directories-and-files which describes something related to the context; and seems that I can add some additional ignores there
* There seems to be new `views` feature for additional includes/excludes but it's unclear that when this is applied

**Comment by helderco** [comment]
@attekei, first of all, the includes/excludes in `dagger.json` affect module loading (*initialize*), while [`DefaultPath` + `Ignore`](https://docs.dagger.io/manuals/developer/functions/#directories-and-files) and views affect passing a `Directory` argument to a function via the CLI (*prepare*). See this for more context:
- https://github.com/dagger/dagger/issues/8542

> My gut feeling is that regardless of strict includes&excludes in dagger.json (btw that json lacks documentation), too much code gets still included.

Sounds like you’re doing it blind (and I get why!), so I think what you need is to inspect the files that were actually uploaded and see which dirs/files weren’t excluded properly.

## Debug function arguments

For `Directory` arguments it’s simple to debug (`DefaultPath` + `Ignore`, a view, or CLI flag without `Ignore` in code). Just create a function that receives a `Directory` as input, and returns the same `Directory`:

```python
@function
def debug(
    self, 
    source: Annotated[
        dagger.Directory,
        DefaultPath("."),
        Ignore([".venv", "sdk"]),
    ],
) -> dagger.Directory:
    return source
```

There’s several things you can do here:
- Calling the function will show you the `Directory`’s digest and top level entries. The digest is content addressed, so it changes if there’s changes in the contents of the directory. Looking at the `entries` field you may be able to spot an interloper:

    ```console
    ❯ dagger call debug
    ```

- You can also list all files, recursively to check it more deeply:
    
    ```console
    ❯ dagger call debug glob --pattern="**/*"
    ```

- Or open the `Directory` in a `Terminal` to navigate around:
    
    ```console
    ❯ dagger call debug terminal
    ```

- Or even export to your host to check it with local tools:
    
    ```console
    ❯ dagger call debug export --path=audit
    ```

Note that these examples assume you’re using `DefaultPath` (i.e., don’t provide it via CLI) because that’s the only time where `Ignore` is used right now, but in the next release you’ll be able to just set `Ignore` and pass different directories via the CLI (`--source` flag in this case):

```console
❯ dagger call debug --source=./subdir
```

<details><summary>Views</summary>

In the meantime, if you can’t wait, views allow providing multiple set of named filters (in `dagger.json`). Assuming a view named “default”, you’d call test it with:

```console
❯ dagger call debug --source=.:default
```

⚠️ However, be warned that views may be deprecated soon in favor of `Ignore` in code.

</details>

## Debug module loading

I’m not sure if there’s a simpler way of doing this, or what that would look like, but I can tell you how I debug it.

Whenever you see a `ModuleSource.resolveFromCaller` step, it’s the engine importing files from your host:

<img width="989" alt="Screenshot 2024-09-28 at 07 37 39" src="https://github.com/user-attachments/assets/0d2f31f5-5589-4994-a3c3-0164a75e98c9">

If it’s taking a long time, it could be that you’re uploading files unnecessarily, and it can possibly have a compounding effect if any changes on unwanted files cause cache misses later on.

When you open it (in the Cloud UI), look for `upload <module context> ... (exclude: ...) (include: ...)`, where “module context” is either the git repo root if the module is in a git repo, or the directory with `dagger.json` if the module isn’t in a git repo).

Notice the **final include/exclude patterns** that the engine is using to filter. It’s not all coming directly from `dagger.json`, that just contributes to it. See if all looks good.

Then, **notice the “blob” step** just after it. *Blobs* are a sort of cache for `Directories` when they’re imported from the host. You can get them through the API but don’t take too long because they can be pruned after a while:

<img width="1122" alt="Screenshot 2024-09-28 at 07 40 43" src="https://github.com/user-attachments/assets/82a8b945-fe07-495f-b2c2-08bb7ef0894b">

For example, using `dagger core`, just copy those inputs:

```console
❯ dagger core blob \
    --digest="sha256: aa44c7c71c77c77cb979724122078b7ecff40d9c6affcb59f9a7ff3fe86175c0" \
    --mediaType="application/vnd.oci.image.layer.v1.tar+zstd" \
    --size=136179 \
    --uncompressed="sha256:cb65d8eb5f603a68afb4be1551b82b32bd381c77907e0fc4651d721a07c0d8ee"
```

The advantage of `dagger core` is you have the same `Directory` printing feature (with `digest` and `entries`), and you can add `terminal` to navigate, but it can be cumbersome to copy the `digest`, `size`, and `uncompressed` inputs multiple times.

The step in the UI is already formatted as a GraphQL query field though, so you can also copy the whole thing and sub-select the field(s) that you’re interested (`digest`, `entries`, `glob`, `export`…):

```console
❯ echo '{ <insert blob() here> { digest glob(pattern: "**/*") } }' | dagger query
```

<details><summary>Full query</summary>

```graphql
{
    blob(digest: "sha256:aa44c7c71c77c77cb979724122078b7ecff40d9c6affcb59f9a7ff3fe86175c0" mediaType: "application/vnd.oci.image.layer.v1.tar+zstd", size: 136179, uncompressed: "sha256:cb65d8eb5f603a68afb4be1551b82b32bd381c77907e0fc4651d721a07c0d8ee") {
        digest
        glob(pattern: "**/*")
    }
}
```

</details>

Or, you can also just copy the `DirectoryID` by clicking the randomly generated icon:

```console
❯ echo '{ loadDirectoryFromID(id: "<insert ID here>") { export(path: "./audit") } }' | dagger query
```

The downside of using a GraphQL query is you may not be familiar with GraphQL enough to figure out the right syntax, but you can create a simple shell script to make this easier, or a function to get `dagger call` capabilities.

<details><summary>Function example</summary>

```python
@function
def load_directory(
    self, 
    from_id: Annotated[string, Doc("DirectoryID value")],
) -> dagger.Directory:
    """Get a Directory object from its ID"""
    return dag.load_directory_from_id(id: dagger.DirectoryID(dir_id))
```

Shell shortcut:
```shell
alias dagdir="dagger call -m github.com/example/daggerverse/debug-utils load-directory --from-id"
```

Example usage:
```console
❯ dagdir DIRECTORYID terminal
```

This is actually a good use case to allow passing ids (in certain cases) to `dagger core`, so watch out for that.

</details>

**Comment by idlsoft** [comment]
I noticed a significant hit if the module in part of a big repository.
The initializer prints something like
```
upload [GIT_ROOT] from ... (client id: ..., session id: ...) (include: [SUBDIR]/dagger.json
```
I did a `git init` in the module directory just to make sure, the time dropped from 40 secs down to 5.
Not sure why it needs to look at the entire repo if `dag.CurrenModule().Source()` only has the subdirectory.

That's with 0.13.7

**Comment by joepio** [comment]
The `connect` and `load module` steps [take **one minute**](https://dagger.cloud/ontola/traces/7375ab5d01a0f0029ba2fac7d33609bf). That's kind of insane. Seems like this could waste a huge amount of Dagger's compute resources, in addition to wasted dev time.

<img width="393" height="156" alt="Image" src="https://github.com/user-attachments/assets/f4c16f2b-77f9-48a1-9f72-74e4cc279115" />

**Comment by eternalphane** [comment]
Just hit the same problem when trying to build dagger itself. Simply running `dagger call cli binary export --help` tooks **~30s each time**, and it seems that subsequent calls to module initialization are not hitting cache.

<img width="3129" height="1247" alt="Image" src="https://github.com/user-attachments/assets/af4a9c45-9125-44fe-89f3-4d0e4983eb84" />
